Completed Chapter 1, handling Top-Level Statements a for initial set of listings.

MarkMichaelis opened the pull request.

MarkMichaelis committed:
Added some test generated directories that should be ignored.

MarkMichaelis committed:
Added top level statement and refactored PowerShellTestsUtilities

MarkMichaelis committed:
Added top level statement and refactored PowerShellTestsUtilities

MarkMichaelis committed:
Added top level statement and refactored PowerShellTestsUtilities

MarkMichaelis committed:
Refactored PowerShellTestsUtilities into a solution shared file.

MarkMichaelis committed:
Refactored PowerShellTestsUtilities into a solution shared file.

MarkMichaelis committed:
Refactored PowerShellTestsUtilities into a solution shared file.

MarkMichaelis committed:
Minor rename of PowerShellTestUtilities

MarkMichaelis committed:
Minor rename of PowerShellTestUtilities

MarkMichaelis committed:
Added a Listing 01.01A as a top level statement.

MarkMichaelis committed:
Excluded "HelloWorld" files 

Also removed top level statement (which wasn't needed in this branch)

MarkMichaelis committed:
Excluded "HelloWorld" files 

Also removed top level statement (which wasn't needed in this branch)

MarkMichaelis committed:
Excluded "HelloWorld" files 

Also removed top level statement (which wasn't needed in this branch)

MarkMichaelis committed:
PowerShell Core by Default

Switched to use PowerShell core by default and to test for PowerShell by executing the PowerShell comment.

MarkMichaelis committed:
Fixed failing tests (from inattention to detail refactoring)

MarkMichaelis committed:
Minor refactoring

MarkMichaelis committed:
Replaced left over debugging line with non-hard-coded path.

MarkMichaelis committed:
Powershell refactoring

MarkMichaelis committed:
Powershell refactoring

MarkMichaelis committed:
Added top level statement

MarkMichaelis committed:
Minor edit

MarkMichaelis committed:
Updated ClassCleanup to use refactored method.

MarkMichaelis committed:
Updated ClassCleanup to use refactored method.

MarkMichaelis committed:
Minor addition of Path.Join()

MarkMichaelis committed:
Added support to only test HelloWorld ps1 execution output when compiling in English

HelloWorld execution output tested regardless.

MarkMichaelis committed:
Started initial draft with top level statement for initial listing

Keboo reviewed (commented):
This looks like it contains a bunch of the changes from #364 can we separate these?

dependabot[bot] committed:
Bump Microsoft.CodeAnalysis.CSharp.Scripting from 4.3.0 to 4.3.1 (#367)

Bumps
[Microsoft.CodeAnalysis.CSharp.Scripting](https://github.com/dotnet/roslyn)
from 4.3.0 to 4.3.1.
<details>
<summary>Release notes</summary>
<p><em>Sourced from <a
href="https://github.com/dotnet/roslyn/releases">Microsoft.CodeAnalysis.CSharp.Scripting's
releases</a>.</em></p>
<blockquote>
<h2>.NET 6.0.1</h2>
<p><a
href="https://github.com/dotnet/core/releases/tag/v6.0.1">Release</a></p>
<h2>.NET 5.0.4</h2>
<p><a
href="https://github.com/dotnet/core/releases/tag/v5.0.4">Release</a></p>
<h2>.NET 5.0.2</h2>
<p><a
href="https://github.com/dotnet/core/blob/master/release-notes/5.0/5.0.2/5.0.2.md">Release
Notes</a>
<a
href="https://github.com/dotnet/core/blob/master/release-notes/5.0/5.0.2/5.0.2-install-instructions.md">Install
Instructions</a></p>
<h1>Repos</h1>
<ul>
<li><a
href="https://github.com/dotnet/core/releases/tag/v5.0.2">Core</a></li>
</ul>
</blockquote>
</details>
<details>
<summary>Commits</summary>
<ul>
<li>See full diff in <a
href="https://github.com/dotnet/roslyn/commits">compare view</a></li>
</ul>
</details>
<br />


[![Dependabot compatibility
score](https://dependabot-badges.githubapp.com/badges/compatibility_score?dependency-name=Microsoft.CodeAnalysis.CSharp.Scripting&package-manager=nuget&previous-version=4.3.0&new-version=4.3.1)](https://docs.github.com/en/github/managing-security-vulnerabilities/about-dependabot-security-updates#about-compatibility-scores)

Dependabot will resolve any conflicts with this PR as long as you don't
alter it yourself. You can also trigger a rebase manually by commenting
`@dependabot rebase`.

[//]: # (dependabot-automerge-start)
[//]: # (dependabot-automerge-end)

---

<details>
<summary>Dependabot commands and options</summary>
<br />

You can trigger Dependabot actions by commenting on this PR:
- `@dependabot rebase` will rebase this PR
- `@dependabot recreate` will recreate this PR, overwriting any edits
that have been made to it
- `@dependabot merge` will merge this PR after your CI passes on it
- `@dependabot squash and merge` will squash and merge this PR after
your CI passes on it
- `@dependabot cancel merge` will cancel a previously requested merge
and block automerging
- `@dependabot reopen` will reopen this PR if it is closed
- `@dependabot close` will close this PR and stop Dependabot recreating
it. You can achieve the same result by closing it manually
- `@dependabot ignore this major version` will close this PR and stop
Dependabot creating any more for this major version (unless you reopen
the PR or upgrade to it yourself)
- `@dependabot ignore this minor version` will close this PR and stop
Dependabot creating any more for this minor version (unless you reopen
the PR or upgrade to it yourself)
- `@dependabot ignore this dependency` will close this PR and stop
Dependabot creating any more for this dependency (unless you reopen the
PR or upgrade to it yourself)


</details>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

MarkMichaelis committed:
Updated for C# 10

MarkMichaelis committed:
Updated to reflect VS new project generated file.

MarkMichaelis committed:
Added comments to identify various constructs

MarkMichaelis committed:
Bumped version numbers to 11.

MarkMichaelis commented:
I have closed PR #364. This PR supersedes it.

Keboo commented on src/Chapter10.Tests/Listing10.23.RegisteringAFinalizerWithProcessExit.Tests.cs:
```diff
@@ -1,41 +1,39 @@
-        static bool PowerShellNotAvailable = PowerShellTestsUtilities.PowerShellNotInstalled();
-        static string PowershellEnvironmentVariableName { get; set; } = "powershell";
+        static string Ps1DirectoryPath { get; } =
+            Path.GetFullPath(Path.Join("..", "..", "..", "..", "Chapter10"), Environment.CurrentDirectory);
```
I typically hate relative paths like this as they make assumptions on the working directory. We have a solution in the other repo that instead of counting a number of parent directories, it instead looks up through the directories for the ".git" folder to find the repo root and builds relative paths based on that.

BenjaminMichaelis replied:
https://www.nuget.org/packages/IntelliTect.Multitool/#readme-body-tab

MarkMichaelis replied:
You are right.  I eliminated the problem in the PowerShell script but not the C# code. @BenjaminMichaelis has been suggesting I use the common library for this but I was reluctant to take an additional library dependency, even in a test. However, your point about working directory is significant.

MarkMichaelis replied:
Leveraged Multitool as @BenjaminMichaelis  suggested.

Keboo commented on src/Shared/Tests/PowerShellTestUtilities.cs~Minor rename of PowerShellTestUtilities:
```diff
@@ -0,0 +1,77 @@
+using System;
```
Did you mean to include this file?

MarkMichaelis replied:
No, but interesting... What created it?

MarkMichaelis committed:
Updated to use IntelliTect.Multitool to find location

MarkMichaelis committed:
Removed errant file

MarkMichaelis committed:
Minor updates

MarkMichaelis committed:
Upgraded to .net7.0/v11

Switched to dotnet 6.0 & 7.0
dotnet required console rather than Console
Include pre-release for setup-dotnet
Update LocalesTest.yml
Updated tests to support different versions of .NET

Updated to C# 11.0
Modified for GitHub build actions.

MarkMichaelis committed:
Merge branch 'v11.0' into TopLevelStatements

MarkMichaelis committed:
Minor update to csproj file

BenjaminMichaelis commented on Directory.Build.props:
```diff
@@ -1,15 +1,16 @@
     <Copyright>Copyright © Addison Wesley 2022</Copyright>
     <Product>Essential C# $(LangVersion)</Product>
     <Title>$(Product)</Title>
-    <Version>9.0</Version>
```
Do we need both a "version" and a "langVersion" or can these be combined or one referenced from the other

MarkMichaelis replied:
Interesting... The history is ambiguous. I'm not seeing where `Version` is used anymore.  When `LangVersion` was added `Version` was kept but I'm not sure why. I have removed `Version`.
Good catch!!

Keboo replied:
Version is used by `dotnet build`, `pack`, and `publish` to set the output version numbers. It, along with a few other properties and the means to control the output version. Given that none of these projects are expected to be done that way  I think it makes sense to omit it.

MarkMichaelis replied:
Thanks... I was pretty sure it wasn't added by mistake but couldn't reproduce the need when I tested it.  Glad to know removing it wasn't a mistake.

BenjaminMichaelis commented on src/Chapter01.Tests/Listing01.01.Tests.cs:
```diff
@@ -1,18 +1,60 @@
+            Path.GetFullPath(Path.Join("..", "..", "..", "..", "Chapter01"), Environment.CurrentDirectory);
+        static string Ps1Path { get; } = 
+            Path.GetFullPath(
+                Path.Join(Ps1DirectoryPath, "Listing01.01.HelloWorldInC#.ps1"), Environment.CurrentDirectory);
```
We could use the MultiTool to normalize the paths here as well

BenjaminMichaelis commented on src/Chapter01.Tests/Listing01.02B.Tests.cs:
```diff
@@ -1,6 +1,6 @@
 ﻿using Microsoft.VisualStudio.TestTools.UnitTesting;
 
-namespace AddisonWesley.Michaelis.EssentialCSharp.Chapter01.Listing01_06.Tests
+namespace AddisonWesley.Michaelis.EssentialCSharp.Chapter01.Listing01_02B.Tests
```
Do we want to merge this in before running the listing manager or wait for the word docs to get updated?

MarkMichaelis replied:
I don't know what the order of operations is.

BenjaminMichaelis commented on src/Chapter01/Listing01.01.HelloWorldInC#.ps1:
```diff
@@ -22,21 +23,24 @@ try {
     Get-Content Program.cs
-    dotnet run
+
+    Write-Warning "Remove dotnet build step once we upgrade to version 10"
+    # Build with version 10 because dotnet version 6 defaults to an ImplicitUsings element in the project file.
+    dotnet build /property:LangVersion=10
+    dotnet run --no-build
```
```suggestion
    dotnet run
```

Aren't we in v11 in this pr now? Should this be updated to remove the build step and remove the warning?

MarkMichaelis replied:
Agreed!

BenjaminMichaelis commented on src/Chapter01/Listing01.04.BreakingApartHelloWorld.cs:
```diff
@@ -1,12 +1,13 @@
 {
     #region INCLUDE
-    public class Program//Class definition
+    public class Program                // BEGIN Class definition
```
We should check how this formats in the manuscript, I wonder what this might look like

MarkMichaelis replied:
I assume it looks like a comment. I'm not sure I understand what you are asking.

BenjaminMichaelis commented on src/Chapter02/Listing02.13.UsingTheNewlineCharacterToInsertANewline.cs:
```diff
@@ -1,14 +1,14 @@
+            #region INCLUDE
             System.Console.Write(
                 "\"Truly, you have a dizzying intellect.\"");
             System.Console.Write("\n\"Wait 'til I get going!\"\n");
```
Do we want to use `Environment.NewLine` here or is it still too early in the book? Or just use Console.WriteLine maybe?
```suggestion
            System.Console.WriteLine("\"Wait 'til I get going!\"");
```

MarkMichaelis replied:
The text is explicit showing the difference between `WriteLine` and `Write`.

BenjaminMichaelis commented on src/Chapter02/Listing02.14B.RawLiteralStrings.cs:
```diff
@@ -0,0 +1,28 @@
+        public static void Main()
+        {
+            #region INCLUDE
+//            System.Console.Write(
+//                """begin
+//           ____
+//          /   /\
+//         /___/  \
+//        / s /\  /\
+//       / u /  \/  \
+//      / i /   /\   \
+//     / b /   /  \   \
+//    / o /   /\   \   \
+//   / m /   /  \   \   \
+//  /___/___/____\   \   \
+// /   /          \   \  /\
+///___/____________\___\/  \
+//\   \             \   \  /
+// \___\_____________\___\/
+//end""");
```
```suggestion
            System.Console.Write(
                """begin
           ____
          /   /\
         /___/  \
        / s /\  /\
       / u /  \/  \
      / i /   /\   \
     / b /   /  \   \
    / o /   /\   \   \
   / m /   /  \   \   \
  /___/___/____\   \   \
 /   /          \   \  /\
/___/____________\___\/  \
\   \             \   \  /
 \___\_____________\___\/
end""");
```
Is this intentionally commented out?

MarkMichaelis replied:
Nope... thanks for noticing.

BenjaminMichaelis commented on src/Chapter10/Listing10.23.RegisteringAFinalizerWithProcessExit.ps1:
```diff
@@ -1,67 +1,79 @@
+            Get-Item $projectDirectory -ErrorAction Ignore | Remove-Item  -Recurse -Force
+            Set-PSDebug -Trace $TraceLevel
+            dotnet new console --output $projectDirectory
+            Set-Content (Join-Path $projectDirectory 'Directory.Build.props') '<Project><PropertyGroup><Nullable>enable</Nullable></PropertyGroup></Project>'
```
Is nullable enabled by default and we don't need to explicitly put that in the proj? However it doesn't hurt so we can just leave it.

BenjaminMichaelis commented on src/Chapter11/Properties/launchSettings.json:
```diff
@@ -0,0 +1,8 @@
+{
+  "profiles": {
+    "WSL": {
+      "commandName": "WSL2",
+      "distributionName": ""
+    }
+  }
+}
```
Do we want to be committing this file?

BenjaminMichaelis commented on src/Shared/Tests/PowerShellTestUtilities.cs:
```diff
@@ -0,0 +1,69 @@
+        {
+            return RuntimeInformation.IsOSPlatform(OSPlatform.Windows);
+        }
+
```
```suggestion
```
nit: extra line

BenjaminMichaelis commented on src/Shared/Tests/PowerShellTestUtilities.cs:
```diff
@@ -0,0 +1,69 @@
+        }
+
+        public static string? PowerShellCommand => _PowerShellCommand.Value;
+        public static bool PowerShellNotInstalled =>_PowerShellCommand.Value == null;
```
```suggestion
        public static bool PowerShellNotInstalled => _PowerShellCommand.Value is null;
```
nit: add a space, and can this be is null?

BenjaminMichaelis reviewed (commented):
I would double check all the warnings your getting are expected

MarkMichaelis committed:
Added Raw Literal String samples

MarkMichaelis committed:
Removed CurrentPath dependency

MarkMichaelis committed:
Updated now that we are on C#11/.NET7.0

MarkMichaelis committed:
Update src/Shared/Tests/PowerShellTestUtilities.cs

Co-authored-by: Benjamin Michaelis <benjamin@michaelis.net>

MarkMichaelis committed:
Removed Version element from Directory.Build.props - relying soley on LangVersion instead

Keboo commented on global.json:
```diff
@@ -1,7 +1,6 @@
-    "version": "6.0.0",
-    "rollForward": "latestMinor",
-    "allowPrerelease": false
+    "rollForward": "latestPatch",
```
Any reason why the version was dropped? This seems problematic as people will not get errors if they don't have 7.0 installed.

Keboo commented on .gitignore:
```diff
@@ -1,5 +1,7 @@
 # Files specific to this project:
 *.encrypt  # Generated by one of the projects and left around on occasion.
+src/Chapter10/ProcessExitTestProgram/
+HelloWorld/
```
Is this expected to cover the HelloWorld program created by the powershell script in chapter 1? If so there is likely a working directory issue. This is filtering from the root directory, while the chapter 1 csproj looks for a HelloWorld directory inside of it. 
I would suggest that we do something to the path's similar to C# files. Perhaps <dir>.g? to indicate that a folder is generated, then we can have a singular globbing pattern. 

`**/*.g/` for this file
`<Compile Remove="**\*.g\**" />` to be put in the Directory.Build.targets file

Keboo commented on src/Chapter01/Chapter01.csproj:
```diff
@@ -8,9 +8,13 @@
   <ItemGroup>
+    <Compile Remove="HelloWorld\**" />
+    <Compile Remove="Listing01.01.HelloWorldInC#.cs" />
+    <Compile Remove="Listing01.01A.HelloWorldInC#.cs" />
```
This one is removed and not added to the `None` set like `Listing01.01.HelloWorldInC#.cs` is this intentional?

MarkMichaelis replied:
@Keboo: Yes... you cannot specify /main (used to run precode via which a user selects which listing to run) with top-level statements. :(.  Alternative suggestions welcome.

MarkMichaelis replied:
Once web edit/run is supported this becomes irrelevant so ignoring.

BenjaminMichaelis committed:
Automerge dependabot and add Github actions to dependabot (#374)

Keboo commented on src/Chapter16.Tests/Listing16.15.Tests.cs:
```diff
@@ -1,22 +1,98 @@
         [TestMethod]
         public void SelectingAnonymousTypeFollowingGroupClause()
         {
+            string e = Shared.NetCore.GetNetCoreVersion() switch
```
This entire switch statement appears unused and duplicated the one below.

MarkMichaelis replied:
Indeed... incomplete refactoring. :)

MarkMichaelis replied:
Indeed... incomplete refactoring. :)

Keboo commented on src/Chapter16.Tests/Listing16.15.Tests.cs:
```diff
@@ -1,22 +1,98 @@
+                    ToList, ToLookup, TryGetNonEnumeratedCount, Union, UnionBy, Where, Zip, 
+                    """;
+
+            TestContext.WriteLine(Shared.NetCore.GetNetCoreVersion());
```
This code appears to invoke `Shared.NetCore.GetNetCoreVersion()` several times. Can we cache the return in a local.
```suggestion
            string dotnetVersion = Shared.NetCore.GetNetCoreVersion(); 
            TestContext.WriteLine(dotnetVersion);
```

Keboo commented on src/Chapter16.Tests/Listing16.15.Tests.cs:
```diff
@@ -1,22 +1,98 @@
+                    """;
+
+            TestContext.WriteLine(Shared.NetCore.GetNetCoreVersion());
+            if (string.Compare(Shared.NetCore.GetNetCoreVersion(), "7") < 0)
```
Is there a reason these are using `string.Compare` rather than `string.StartsWith`?

dependabot[bot] committed:
Bump actions/upload-artifact from 2 to 3 (#381)

Bumps
[actions/upload-artifact](https://github.com/actions/upload-artifact)
from 2 to 3.
<details>
<summary>Release notes</summary>
<p><em>Sourced from <a
href="https://github.com/actions/upload-artifact/releases">actions/upload-artifact's
releases</a>.</em></p>
<blockquote>
<h2>v3.0.0</h2>
<h2>What's Changed</h2>
<ul>
<li>Update default runtime to node16 (<a
href="https://github-redirect.dependabot.com/actions/upload-artifact/issues/293">#293</a>)</li>
<li>Update package-lock.json file version to 2 (<a
href="https://github-redirect.dependabot.com/actions/upload-artifact/issues/302">#302</a>)</li>
</ul>
<h3>Breaking Changes</h3>
<p>With the update to Node 16, all scripts will now be run with Node 16
rather than Node 12.</p>
<h2>v2.3.1</h2>
<p>Fix for empty fails on Windows failing on upload <a
href="https://github-redirect.dependabot.com/actions/upload-artifact/issues/281">#281</a></p>
<h2>v2.3.0 Upload Artifact</h2>
<ul>
<li>Optimizations for faster uploads of larger files that are already
compressed</li>
<li>Significantly improved logging when there are chunked uploads</li>
<li>Clarifications in logs around the upload size and prohibited
characters that aren't allowed in the artifact name or any uploaded
files</li>
<li>Various other small bugfixes &amp; optimizations</li>
</ul>
<h2>v2.2.4</h2>
<ul>
<li>Retry on HTTP 500 responses from the service</li>
</ul>
<h2>v2.2.3</h2>
<ul>
<li>Fixes for proxy related issues</li>
</ul>
<h2>v2.2.2</h2>
<ul>
<li>Improved retryability and error handling</li>
</ul>
<h2>v2.2.1</h2>
<ul>
<li>Update used actions/core package to the latest version</li>
</ul>
<h2>v2.2.0</h2>
<ul>
<li>Support for artifact retention</li>
</ul>
<h2>v2.1.4</h2>
<ul>
<li>Add Third Party License Information</li>
</ul>
<h2>v2.1.3</h2>
<ul>
<li>Use updated version of the <code>@action/artifact</code> NPM
package</li>
</ul>
<h2>v2.1.2</h2>
<ul>
<li>Increase upload chunk size from 4MB to 8MB</li>
<li>Detect case insensitive file uploads</li>
</ul>
<h2>v2.1.1</h2>
<ul>
<li>Fix for certain symlinks not correctly being identified as
directories before starting uploads</li>
</ul>
<h2>v2.1.0</h2>
<ul>
<li>Support for uploading artifacts with multiple paths</li>
<li>Support for using exclude paths</li>
<li>Updates to dependencies</li>
</ul>
<!-- raw HTML omitted -->
</blockquote>
<p>... (truncated)</p>
</details>
<details>
<summary>Commits</summary>
<ul>
<li><a
href="https://github.com/actions/upload-artifact/commit/83fd05a356d7e2593de66fc9913b3002723633cb"><code>83fd05a</code></a>
Bump actions-core to v1.10.0 (<a
href="https://github-redirect.dependabot.com/actions/upload-artifact/issues/356">#356</a>)</li>
<li><a
href="https://github.com/actions/upload-artifact/commit/3cea5372237819ed00197afe530f5a7ea3e805c8"><code>3cea537</code></a>
Merge pull request <a
href="https://github-redirect.dependabot.com/actions/upload-artifact/issues/327">#327</a>
from actions/robherley/artifact-1.1.0</li>
<li><a
href="https://github.com/actions/upload-artifact/commit/849aa7758a428ee22be38de371b49c8bd57c4b9d"><code>849aa77</code></a>
nvm use 12 &amp; npm run release</li>
<li><a
href="https://github.com/actions/upload-artifact/commit/4d3986961d0423ba9a593efb490a2960eb65f43b"><code>4d39869</code></a>
recompile with correct ncc version</li>
<li><a
href="https://github.com/actions/upload-artifact/commit/2e0d362ec5cf81e334dda822c49c96dcd4b7df2c"><code>2e0d362</code></a>
bump <code>@​actions/artifact</code> to 1.1.0</li>
<li><a
href="https://github.com/actions/upload-artifact/commit/09a5d6a283da3e7c9f3253a5d4cdab2347712a66"><code>09a5d6a</code></a>
Merge pull request <a
href="https://github-redirect.dependabot.com/actions/upload-artifact/issues/320">#320</a>
from actions/dependabot/npm_and_yarn/ansi-regex-4.1.1</li>
<li><a
href="https://github.com/actions/upload-artifact/commit/189315d9106f43a2a8eb60836608bb96b1f69d77"><code>189315d</code></a>
Bump ansi-regex from 4.1.0 to 4.1.1</li>
<li><a
href="https://github.com/actions/upload-artifact/commit/d159c2d80bf32e77611286e4d71bfe6d15208d88"><code>d159c2d</code></a>
Merge pull request <a
href="https://github-redirect.dependabot.com/actions/upload-artifact/issues/297">#297</a>
from actions/dependabot/npm_and_yarn/ajv-6.12.6</li>
<li><a
href="https://github.com/actions/upload-artifact/commit/c26a7ba4b5dbaecea906fec3b7d2c0c86f1ccaba"><code>c26a7ba</code></a>
Bump ajv from 6.11.0 to 6.12.6</li>
<li><a
href="https://github.com/actions/upload-artifact/commit/6ed6c729229a623bcb1fdd75903dc6e436b3d0a7"><code>6ed6c72</code></a>
Merge pull request <a
href="https://github-redirect.dependabot.com/actions/upload-artifact/issues/303">#303</a>
from actions/dependabot/npm_and_yarn/yargs-parser-13.1.2</li>
<li>Additional commits viewable in <a
href="https://github.com/actions/upload-artifact/compare/v2...v3">compare
view</a></li>
</ul>
</details>
<br />


[![Dependabot compatibility
score](https://dependabot-badges.githubapp.com/badges/compatibility_score?dependency-name=actions/upload-artifact&package-manager=github_actions&previous-version=2&new-version=3)](https://docs.github.com/en/github/managing-security-vulnerabilities/about-dependabot-security-updates#about-compatibility-scores)

Dependabot will resolve any conflicts with this PR as long as you don't
alter it yourself. You can also trigger a rebase manually by commenting
`@dependabot rebase`.

[//]: # (dependabot-automerge-start)
[//]: # (dependabot-automerge-end)

---

<details>
<summary>Dependabot commands and options</summary>
<br />

You can trigger Dependabot actions by commenting on this PR:
- `@dependabot rebase` will rebase this PR
- `@dependabot recreate` will recreate this PR, overwriting any edits
that have been made to it
- `@dependabot merge` will merge this PR after your CI passes on it
- `@dependabot squash and merge` will squash and merge this PR after
your CI passes on it
- `@dependabot cancel merge` will cancel a previously requested merge
and block automerging
- `@dependabot reopen` will reopen this PR if it is closed
- `@dependabot close` will close this PR and stop Dependabot recreating
it. You can achieve the same result by closing it manually
- `@dependabot ignore this major version` will close this PR and stop
Dependabot creating any more for this major version (unless you reopen
the PR or upgrade to it yourself)
- `@dependabot ignore this minor version` will close this PR and stop
Dependabot creating any more for this minor version (unless you reopen
the PR or upgrade to it yourself)
- `@dependabot ignore this dependency` will close this PR and stop
Dependabot creating any more for this dependency (unless you reopen the
PR or upgrade to it yourself)


</details>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

Keboo commented on .github/workflows/build-and-test.yml:
```diff
@@ -19,12 +19,12 @@ jobs:
       - uses: actions/checkout@v3
       - name: Setup dotnet SDK
-        uses: actions/setup-dotnet@v2
+        uses: actions/setup-dotnet@v3
```
dependabot update has just merged that will keep these up to date.

BenjaminMichaelis replied:
Might of been me when I was helping fix the action on his computer for 7.0 that I bumped it at the same time...

dependabot[bot] committed:
Bump actions/setup-dotnet from 2 to 3 (#380)

Bumps [actions/setup-dotnet](https://github.com/actions/setup-dotnet)
from 2 to 3.
<details>
<summary>Release notes</summary>
<p><em>Sourced from <a
href="https://github.com/actions/setup-dotnet/releases">actions/setup-dotnet's
releases</a>.</em></p>
<blockquote>
<h2>v3.0.0</h2>
<p>This major release includes the following
<strong>changes:</strong></p>
<ul>
<li><a
href="https://github-redirect.dependabot.com/actions/setup-dotnet/issues/219">#219</a>
New input <code>dotnet-quality</code> was added in <a
href="https://github-redirect.dependabot.com/actions/setup-dotnet/pull/315">#315</a>:</li>
</ul>
<pre lang="yaml"><code>    - uses: actions/setup-dotnet@v3
      with:
        dotnet-version: '6.0.x'
        dotnet-quality: 'preview'
    - run: dotnet build &lt;my project&gt;
</code></pre>
<p>More in detail <a
href="https://github.com/actions/setup-dotnet#using-the-dotnet-quality-input">here</a>.</p>
<ul>
<li><a
href="https://github-redirect.dependabot.com/actions/setup-dotnet/issues/241">#241</a>
The output variable <code>dotnet-version</code> which contains the
installed by the action SDK version was added in <a
href="https://github-redirect.dependabot.com/actions/setup-dotnet/pull/324">#324</a>:</li>
</ul>
<pre lang="yaml"><code>    - uses: actions/setup-dotnet@v3
      id: cp310
      with:
        dotnet-version: '3.1.422'
- run: echo '${{ steps.cp310.outputs.dotnet-version }}' # outputs
3.1.422
</code></pre>
<p>More in detail <a
href="https://github.com/actions/setup-dotnet/tree/main#dotnet-version">here</a>.</p>
<ul>
<li>The <code>dotnet-version</code> syntax was updated and now it allows
to specify the prerelease version without using
<code>include-prerelease</code> input. The
<code>include-prerelease</code> input was cut out:</li>
</ul>
<pre lang="yaml"><code>    - uses: actions/setup-dotnet@v3
      with:
        dotnet-version: '5.0.0-preview.6'
</code></pre>
<p>More in detail <a
href="https://github.com/actions/setup-dotnet#supported-version-syntax">here</a>.</p>
<ul>
<li><a
href="https://github-redirect.dependabot.com/actions/setup-dotnet/issues/251">#251</a>
The problem with out of support .NET version warnings was solved in <a
href="https://github-redirect.dependabot.com/actions/setup-dotnet/pull/315">#315</a>.</li>
</ul>
<p><strong>Breaking changes</strong>:</p>
<ul>
<li>Installation paths for Windows and Ubuntu images were changed to
match the location of pre-installed SDKs. In more detail, read <a
href="https://github.com/actions/setup-dotnet/blob/main/docs/adrs/v3-setup-dotnet.md#breaking-changes">here</a>.</li>
</ul>
<h2>Add support for Windows-arm</h2>
<p>In scope of this release we <a
href="https://github-redirect.dependabot.com/actions/setup-dotnet/pull/320">add
support for Windows-arm</a>. Besides, we change getInput to <a
href="https://github-redirect.dependabot.com/actions/setup-dotnet/pull/250">getBooleanInput</a>
for include-prerelease.</p>
<h2>Package updates, support for global json file in a subdirectory,
installer scripts updates</h2>
<p>This release includes the following PRs:</p>
<ul>
<li>Adding support for the <code>global-json-file</code> input: <a
href="https://github-redirect.dependabot.com/actions/setup-dotnet/issues/276">#276</a>
Example of usage:
<pre lang="yaml"><code>- uses: actions/setup-dotnet@v2
  with:
    global-json-file: csharp/global.json
- run: dotnet build &lt;my project&gt;
  working-directory: csharp
</code></pre>
</li>
</ul>
<!-- raw HTML omitted -->
</blockquote>
<p>... (truncated)</p>
</details>
<details>
<summary>Commits</summary>
<ul>
<li><a
href="https://github.com/actions/setup-dotnet/commit/4d4a70f4a5b2a5a5329f13be4ac933f2c9206ac0"><code>4d4a70f</code></a>
Add logic for processing DOTNET_INSTALL_DIR environment variable (<a
href="https://github-redirect.dependabot.com/actions/setup-dotnet/issues/332">#332</a>)</li>
<li><a
href="https://github.com/actions/setup-dotnet/commit/70c3f4d09839515e1a5435d1ef8962637486aad1"><code>70c3f4d</code></a>
Merge pull request <a
href="https://github-redirect.dependabot.com/actions/setup-dotnet/issues/331">#331</a>
from rentziass/rentziass/update-actions-core</li>
<li><a
href="https://github.com/actions/setup-dotnet/commit/92606438168a6a25f2939b37f03d67816d27b1c5"><code>9260643</code></a>
Update <code>@​actions/core</code> to 1.10.0</li>
<li><a
href="https://github.com/actions/setup-dotnet/commit/45c9f236cf9a0c09d11298348f6ed06bd81914c0"><code>45c9f23</code></a>
Enable ability to change .NET SDK installation directory by
`DOTNET_INSTALL_D...</li>
<li><a
href="https://github.com/actions/setup-dotnet/commit/c7e7147fd3e41056e75a761416c06f5fa781b5fb"><code>c7e7147</code></a>
Add ability to write resolved version of SDK into the output variable
(<a
href="https://github-redirect.dependabot.com/actions/setup-dotnet/issues/324">#324</a>)</li>
<li><a
href="https://github.com/actions/setup-dotnet/commit/0705ef0281736c0ca177266c8d04eb08f72072de"><code>0705ef0</code></a>
Implement proposal stated in ADR for setup-dotnet v3 and functionality
from f...</li>
<li>See full diff in <a
href="https://github.com/actions/setup-dotnet/compare/v2...v3">compare
view</a></li>
</ul>
</details>
<br />


[![Dependabot compatibility
score](https://dependabot-badges.githubapp.com/badges/compatibility_score?dependency-name=actions/setup-dotnet&package-manager=github_actions&previous-version=2&new-version=3)](https://docs.github.com/en/github/managing-security-vulnerabilities/about-dependabot-security-updates#about-compatibility-scores)

Dependabot will resolve any conflicts with this PR as long as you don't
alter it yourself. You can also trigger a rebase manually by commenting
`@dependabot rebase`.

[//]: # (dependabot-automerge-start)
[//]: # (dependabot-automerge-end)

---

<details>
<summary>Dependabot commands and options</summary>
<br />

You can trigger Dependabot actions by commenting on this PR:
- `@dependabot rebase` will rebase this PR
- `@dependabot recreate` will recreate this PR, overwriting any edits
that have been made to it
- `@dependabot merge` will merge this PR after your CI passes on it
- `@dependabot squash and merge` will squash and merge this PR after
your CI passes on it
- `@dependabot cancel merge` will cancel a previously requested merge
and block automerging
- `@dependabot reopen` will reopen this PR if it is closed
- `@dependabot close` will close this PR and stop Dependabot recreating
it. You can achieve the same result by closing it manually
- `@dependabot ignore this major version` will close this PR and stop
Dependabot creating any more for this major version (unless you reopen
the PR or upgrade to it yourself)
- `@dependabot ignore this minor version` will close this PR and stop
Dependabot creating any more for this minor version (unless you reopen
the PR or upgrade to it yourself)
- `@dependabot ignore this dependency` will close this PR and stop
Dependabot creating any more for this dependency (unless you reopen the
PR or upgrade to it yourself)


</details>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

MarkMichaelis committed:
Updated to use .NET 6 or higher

MarkMichaelis committed:
Change extension on test/generated folders to "*.g"

MarkMichaelis committed:
Removed System. prefix.

And fixed line breaks for 1.19 comments example.

MarkMichaelis committed:
Removed stale code

MarkMichaelis committed:
Renamed HelloWorld class to Program

Keboo commented on src/Chapter01/Listing01.08.NoIndentationFormatting.cs:
```diff
@@ -5,7 +5,7 @@ public class HelloWorld
 public static void Main()
 {
-System.Console.WriteLine("Hello Inigo Montoya");
+            Console.WriteLine("Hello Inigo Montoya");
```
Given this is the no indentation example, do we really want this indentation?

Keboo commented on src/Chapter01/Listing01.09.RemovingWhitespace.cs:
```diff
@@ -2,6 +2,6 @@ namespace AddisonWesley.Michaelis.EssentialCSharp.Chapter01.Listing01_09
     #region INCLUDE
     public class HelloWorld{public static void Main()
-{System.Console.WriteLine("Hello Inigo Montoya");}}
+{ Console.WriteLine("Hello Inigo Montoya");}}
```
Inconsistent spacing and too many closing braces
```suggestion
{ Console.WriteLine("Hello Inigo Montoya"); }
```

Keboo commented on src/Chapter10.Tests/Listing10.23.RegisteringAFinalizerWithProcessExit.Tests.cs:
```diff
@@ -1,41 +1,43 @@
-            string testStatus = "cleanup";
-            Process powershell = Process.Start(PowershellEnvironmentVariableName, $"-noprofile -command \"{Ps1Path} 0 null {testStatus}\"");
-            powershell.WaitForExit();
+            Assert.AreEqual<int>(0, RunPowerShellScript(
+                "cleanup", out string _));
```
This assertion doesn't actually work.
By the time the cleanup method gets invoked the all the tests in the class have already reported success/failure.

![image](https://user-images.githubusercontent.com/952248/197562138-3a1670fc-eb22-4818-937a-82d928a9115f.png)

MarkMichaelis replied:
Agreed... Also, if debugging and you stop, cleanup won't run.  Removing the assert and adding cleanup to classinitialize as well.

Keboo commented on .gitignore:
```diff
@@ -1,5 +1,7 @@
 # Files specific to this project:
 *.encrypt  # Generated by one of the projects and left around on occasion.
+**/*.gc/ # Folders generated buy tests will use the *.g extension so they aren't check in.
```
```suggestion
**/*.g/ # Folders generated buy tests will use the *.g extension so they aren't check in.
```

MarkMichaelis committed:
Removed errant indent and whitespace

MarkMichaelis committed:
Added cleanup at start (in addition to end)

MarkMichaelis committed:
Change extension on test/generated folders to "*.testing"

MarkMichaelis committed:
Merge branch 'v9.0' into TopLevelStatements

BenjaminMichaelis commented on src/Chapter01.Tests/Listing01.01.Tests.cs:
```diff
@@ -1,18 +1,61 @@
+                foreach (string text in expectedText)
+                {
+                    Assert.IsTrue(psOutput.Contains(text),
+                        $"Unexpectedly, '{text}' did not appear in the console output: \n {psOutput}");
```
```suggestion
                        $"Unexpectedly, '{text}' did not appear in the console output: {Environment.NewLine} {psOutput}");
```

MarkMichaelis committed:
Minor update

Co-authored-by: Benjamin Michaelis <benjamin@michaelis.net>
